Merge RK schemes into pecos-dev

trevilo opened the pull request:
This code adds three Runge-Kutta time marching schemes for use with the flow equations.  So, it addresses issue #4.

trevilo committed:
Fixing explicit RK implementation.  Coeffs hardcoded to classical RK4 for the moment.

trevilo committed:
Add (in comments) coeffs for Huen's method.

trevilo committed:
Beginning work on IMEX by making giving it a method in CSolver and
associated flow control in CIntegration.  No schemes implemented yet
though... just dies with an error.

trevilo committed:
a dirty implementation of smr91

trevilo committed:
Clean up explicit RK input options

trevilo committed:
Changing naming convention for linearized imex to be more descriptive and make room for two classes of scheme.

trevilo committed:
Minor clean up in SMR91 comments

trevilo committed:
dirty implementation of linearly implicit w/ EDIRK

trevilo committed:
read implicit RK coefficients for linearized imex w/ EDIRK

trevilo committed:
Begin mods necessary to only eval the Jacobian for first RK step.  Lots still to go... mainly boundary condition routines.

trevilo committed:
Slowly working through adding rk step as input to BC fcns.

trevilo committed:
More mods to make room for rk step input in BC fcns

trevilo committed:
All bc fcns in CEulerSolver now take iRKStep

trevilo committed:
Make sure Euler/NS BCs do right thing wrt Jacobian on rk substeps

trevilo committed:
Eliminate Jacobian copying in LIMEX

trevilo committed:
clearing FIXMEs

trevilo committed:
Eliminate some dead code from comments

trevilo committed:
Bug fix for SMR91.  Be sure to set Delta U to zero on first RK substep.

trevilo committed:
Merge branch 'pecos-dev' into pecos-rk

trevilo committed:
Bug fixes.  These fix the navierstokes/cylinder test case.

trevilo committed:
Remove deprecated option RK_ALPHA_COEFFS from all example config files

trevilo committed:
Add RK4 coefficients (to replace deprecated RK_ALPHA_COEFFS option).

trevilo committed:
Update expected results for 'channel' case b/c of fix to RK scheme.

trevilo committed:
rm unnecessary print statements

trevilo committed:
clean up commented out code

trevilo committed:
clean up dead code in comments

trevilo committed:
clean up dead code in comments

trevilo committed:
fix 'TODO' re: iRKStep

trevilo committed:
clean up commented out print

trevilo committed:
Eliminate unnecessary variable Jacobian0

clarkpede commented:
So why is the euler channel RK case failing? Is it expected to fail?

clarkpede commented on SU2_CFD/src/solver_direct_mean.cpp:
```diff
@@ -6438,6 +6510,297 @@ void CEulerSolver::ImplicitEuler_Iteration(CGeometry *geometry, CSolver **solver
+  dt = node[0]->GetDelta_Time();
+
+  // SMR91 coefficients are hardcoded for now
+  // TODO: Let user set these through the input file
```
Are we planning on letting the user input these? Or do we want to keep them hardcoded?

trevilo replied:
Eventually, yes.  For now I vote we leave it as it isn't blocking anything.

clarkpede commented on SU2_CFD/src/solver_direct_mean.cpp:
```diff
@@ -6438,6 +6510,297 @@ void CEulerSolver::ImplicitEuler_Iteration(CGeometry *geometry, CSolver **solver
+  // simulations, such that dt is the same for each node and we can
+  // just grab it off the first node.
+  assert(config->GetUnsteady_Simulation() == TIME_STEPPING);
+  dt = node[0]->GetDelta_Time();
```
This is a bit hacky.  The proper way of doing this would be to pull the Delta_Time from config, but SU2 doesn't update config with the unsteady time for `TIME_STEPPING`.  We should fix the problems with time-stepping, then come back and change this.

trevilo replied:
yeah, I agree it is a bit ugly.  But, works fine for now.  I don't see much motivation to change it until we do a bigger clean up of how time and dt are handled for time stepping.

clarkpede commented on TestCases/fea_fsi/DynBeam_2d/configBeam_2d.cfg:
```diff
@@ -27,7 +27,7 @@ DYN_TIMESTEP= 0.01
 EXT_ITER=2
 TIME_DISCRE_FEA= NEWMARK_IMPLICIT
-NEWMARK_ALPHA=0.2601
+NEWMA
```
Was this an intentional change?

trevilo replied:
Nope... a result of my sed script matching on RK_ALPHA.  Doh.  Good catch.  Fixed in 
8a9f4a5

trevilo committed:
Revise expected results for channel case in serial regression tests b/c different scheme being tested

trevilo commented:
The euler channel RK case expected results changed b/c I changed the RK scheme it is testing.  The old RK implementation only supported a very limited version of RK and the coefficients used led to a 1st order accurate method.  The new implementation is much more general and the test now uses a higher-order accurate scheme (classic RK4).  We could have it use the old scheme, but I don't really see the point.

Anyway, I had updated the results in the parallel version of the tests but forgot the serial.  97f532d fixes the expected results in the serial version as well.

clarkpede commented on TestCases/harmonic_balance/HB.cfg:
```diff
@@ -121,8 +121,8 @@ CFL_ADAPT= NO
 %
-% Runge-Kutta alpha coefficients
-RK_ALPHA_COEFF= ( 0.66667, 0.66667, 1.000000 )
+
```
We should remove these empty lines from the test files.
EDIT: And the blank line with `%` above or below as well (so we don't end up with two `%` rows)

trevilo replied:
White space clean up in 
3c520f1 and 42aa125

clarkpede commented:
I guessed that it was something like that.  I just wanted to make sure.

Were you planning to support multigrid with these three RK schemes? Or is that not implemented?

trevilo commented:
No multigrid support at the moment for the IMEX schemes b/c they `assert(config->GetUnsteady_Simulation() == TIME_STEPPING)` and multigrid is disabled in this mode.

In theory, the explicit RK implementation supports multigrid.  The euler/channel regression test uses it, so it at least runs.  I've confirmed that if I use RK coefficients to reproduce the old 3 stage, 1st order scheme, I get the same results from before, and the new RK4 results are similar.  So, as far as I can tell, it is working, but this has not been extensively investigated.

clarkpede commented on SU2_CFD/src/solver_direct_mean.cpp:
```diff
@@ -6250,36 +6294,63 @@ void CEulerSolver::TurboPerformance(CSolver *solver, CConfig *config, unsigned s
+
+  const bool final_step = (iRKStep+1 == nRKStep);
+
+  // FIXME: Not currently updating time!  Do it!
```
What's the status on this FIXME? Does it need to be resolved, or are we fine leaving it how it is?

trevilo replied:
I vote we leave it for now as it is not yet blocking anything.

clarkpede committed:
Temporarily disabling parallel AD tests until we find a way to cut down compile time

clarkpede commented on SU2_CFD/src/solver_direct_mean.cpp:
```diff
@@ -6438,6 +6510,297 @@ void CEulerSolver::ImplicitEuler_Iteration(CGeometry *geometry, CSolver **solver
+
+  bool adjoint = config->GetContinuous_Adjoint();
+
+  const bool final_step = (iRKStep+1 == nRKStep);
```
`final_step` isn't used.  Should we be using it to do something different the last step? Or should it be deleted?

clarkpede committed:
Removing some unused variables

clarkpede commented on Common/src/config_structure.cpp:
```diff
@@ -4341,11 +4433,67 @@ void CConfig::SetOutput(unsigned short val_software, unsigned short val_izone) {
           cout << "Number of steps: " << nRKStep << endl;
-          cout << "Alpha coefficients: ";
+          cout << "RK coefficients: " << endl;
+          cout << "  Amat = " << endl;
```
I propose changing these "Amat", "bVec" printouts to more readable "A matrix:", "b Vector", etc.  What do you think?

trevilo replied:
Fixed in 
e20ff97

clarkpede commented on Common/src/config_structure.cpp:
```diff
@@ -4341,11 +4433,67 @@ void CConfig::SetOutput(unsigned short val_software, unsigned short val_izone) {
+          for (unsigned short iRKStep = 0; iRKStep < nRKStep-1; iRKStep++) {
+            cout << " " << RK_cVec_imp[iRKStep];
+          }
+          cout << "]" << endl;
```
You're missing `RUNGE_KUTTA_LIMEX_SMR91` from this switch/case.  Was that intentional?

trevilo replied:
Fixed in 
434e685

clarkpede commented:
I just pushed two changes:

1. I disabled the parallel AD tests temporarily.  They were taking too long to compile and maxing out the time limit for CI tests. I believe getting no results is better than a false negative, especially since serial AD is still working.
2.  I also removed some unused variables.

I've got some other feedback too:

+ The config file in the project root, `config_template.cfg`, should be updated with the new settings.  It still has the old `RK_ALPHA_COEFF`
+ I noticed that if you turn on any of the RK schemes, you _must_ specify the number of steps and the coefficients.  Otherwise, it defaults to 0 RK steps and no coefficients. This seems like a bad idea from both a usability and an error-prevention standpoint.  If none of the options are specified (`N_RK_STEP`, `RK_AMAT_LOWER`, etc.) then I believe the program should default to a set of hardcoded values. This makes the new RK schemes easier to use, since you don't have to specify the coefficients every time.  Using the hardcoded values also prevents errors, since you don't have to worry about accidentally putting in the wrong coefficients when all you want are the standard values.
+ I also tried running a steady-state problem with multigrid and explicit RK.  I got a segfault while it was calculating the first step.  We don't need steady-state to work with multigrid and explicit RK, so we could just fence off that possibility.

trevilo commented:
Thanks Clark.  Regarding the template config file, yes, I agree.  For the defaults, yeah, that makes sense.  I'll make these fixes asap.

Can you give details on the case you're running that seg faults?  I don't mind making steady-state+multigrid+RK illegal since we're not going to use it.  But, I'm surprised it is seg faulting b/c this option is used by the euler/channel regression test, which runs fine.

trevilo committed:
Add example RK coefficients to template configuration

clarkpede commented:
I used the flat plate boundary layer case with the SA model. I attached a grid and config file that give you the error.

I discovered that the segfault disappears if you switch the flow numerics from `2ND_ORDER` to `2ND_ORDER_LIMITER`. That explains why the euler channel regression test isn't failing.

[rk-mgrid-segfault.tar.gz](https://github.com/pecos-hybrid/SU2/files/1422475/rk-mgrid-segfault.tar.gz)

trevilo committed:
Make RK4 the default explicit RK if no coeffs in config file.

trevilo committed:
Add protection to avoid write to jacobian in RoeSoundSpeed2<=0 case
when using explicit time marching.  Should fix the seg fault observed
by @clarkpede on #6.

trevilo commented:
Hey Clark, thanks for the comments.  I'll address asap.

In the meantime, a9c18ca ought to fix the seg fault you observed.  There was a spot in the Roe upwind flux calc where the Jacobian was being set regardless of whether running implicit or explicit in time.  It was only accessed when a non-physical condition was encountered, which explains the differences observed for different test cases and spatial schemes.

trevilo committed:
Add default coefficients for LIMEX EDIRK

trevilo committed:
Print message when SMR91 detected.

trevilo committed:
Make RK coefficient output more readable.

trevilo committed:
whitespace cleanup required by naive deletion of deprecated RK_ALPHA_COEFFS option

trevilo committed:
fix NEWMARK_ALPHA option bug caused by naive deletion of RK_ALPHA_COEFFS

trevilo committed:
cleanup required b/c naive deletion of deprecated RK_ALPHA_COEFFS option
